src/main/java/com/minecolonies/core/entity/pathfinding/pathjobs/AbstractPathJob.java

```diff
@@ -276,71 +303,236 @@ protected Path search()
+                    recalcHeuristic(bestNode);
+                    recalcHeuristic(node);
+                }
+            }
+
+            if (!node.isVisited() && !node.isCornerNode())
+            {
+                if (node.getHeuristic() < heuristicEndNodeCutoff || totalNodesVisited > maxNodes * 0.7)
                 {
                     // Calculates a score for a possible end node, defaults to heuristic(closest)
                     final double nodeEndSCore = getEndNodeScore(node);
                     if (nodeEndSCore < bestNodeEndScore)
                     {
                         if (!reachesDestination || isAtDestination(node))
                         {
+                            nodesSinceEndNode = 0;
                             bestNode = node;
                             bestNodeEndScore = nodeEndSCore;
                         }
                     }
                 }
-                else
+                else if (node.getHeuristic() < bestNode.getHeuristic())
                 {
+                    nodesSinceEndNode = 0;
                     bestNode = node;
                     bestNodeEndScore = node.getHeuristic();
                 }
             }
 
+            // Don't keep searching more costly nodes when there is a destination
+            if (!node.isVisited() && reachesDestination && node.getScore() > bestNode.getScore() + 2)
+            {
+                if (reevaluteHeuristic(bestNode, reachesDestination))
+                {
+                    heuristicEndNodeCutoff = startNode.getHeuristic() > 0 ? startNode.getHeuristic() / 3 : 200;
+                    recalcHeuristic(bestNode);
+                    recalcHeuristic(node);
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            handleDebugOptions(node);
             visitNode(node);
+            node.setVisited();
+        }
+
+        // Explore additional possible endnodes after reaching, if we got extra nodes to search
+        if (extraNodes > 0 && reachesDestination)
+        {
+            // Make sure to expand from the final node
+            visitNode(bestNode);
+
+            // Search only closest nodes to the goal
+            final Queue<MNode> original = nodesToVisit;
+            nodesToVisit = new PriorityQueue<>(nodesToVisit.size(), (a, b) -> {
+                if ((a.getHeuristic()) < (b.getHeuristic()))
+                {
+                    return -1;
+                }
+                else if (a.getHeuristic() > b.getHeuristic())
+                {
+                    return 1;
+                }
+                else
+                {
+                    return a.getCounterAdded() - b.getCounterAdded();
+                }
+            });
+            nodesToVisit.addAll(original);
+
+            while (!nodesToVisit.isEmpty())
+            {
+                if (Thread.currentThread().isInterrupted())
```

Comment:
Why this interrupted check here? and why not return finalize?

Reply:
I just copied the check from the main loop :D Afaik this is only triggered when the executor is shutting down, aka game is stopping so returning a path doesnt matter

Reply:
ah okay, it's just to avoid preventing shutdown with this

Reply:
ye, not sure if its really needed though but w/e